src/group.ts

```diff
@@ -339,50 +346,58 @@ export class Group {
-        tv2 = gxd.mul(B) //             16. tv2 = B * gxd
-        gx1 = gx1.add(tv2) //           17. gx1 = gx1 + tv2
-        tv4 = gxd.square() //           18. tv4 = gxd^2
-        tv2 = gx1.mul(gxd) //           19. tv2 = gx1 * gxd
-        tv4 = tv4.mul(tv2) //           20. tv4 = tv4 * tv2
-        let y1 = tv4.power(c1) //       21.  y1 = tv4^c1
-        y1 = y1.mul(tv2) //             22.  y1 = y1 * tv2
-        const x2n = tv3.mul(x1n) //     23. x2n = tv3 * x1n
-        let y2 = y1.mul(c2) //          24.  y2 = y1 * c2
-        y2 = y2.mul(tv1) //             25.  y2 = y2 * tv1
-        y2 = y2.mul(u) //               26.  y2 = y2 * u
-        tv2 = y1.square() //            27. tv2 = y1^2
-        tv2 = tv2.mul(gxd) //           28. tv2 = tv2 * gxd
-        const e2 = tv2.equals(gx1) //  29.  e2 = tv2 == gx1
-        const xn = cmov(x2n, x1n, e2) //  30.  xn = CMOV(x2n, x1n, e2)
-        let y = cmov(y2, y1, e2) //     31.   y = CMOV(y2, y1, e2)
-        const e3 = sgn(u) === sgn(y) // 32.  e3 = sgn0(u) == sgn0(y)
-        tv1 = p.sub(y)
-        y = cmov(tv1, y, e3) //         33.   y = CMOV(-y, y, e3)
-        let x = xd.inverseMod(p) //     34. return (xn, xd, y, 1)
-        x = xn.mul(x)
-
-        const point = new sjcl.ecc.point(this.curve, new sjcl.bn(x), new sjcl.bn(y))
+        let tv1 = u.square() //         1.  tv1 = u^2
+        tv1 = Z.mul(tv1) //             2.  tv1 = Z * tv1
+        let tv2 = tv1.square() //       3.  tv2 = tv1^2
+        tv2 = tv2.add(tv1) //           4.  tv2 = tv2 + tv1
+        let tv3 = tv2.add(one) //       5.  tv3 = tv2 + 1
+        tv3 = B.mul(tv3) //             6.  tv3 = B * tv3
+        let tv4 = cmov(Z, zero.sub(tv2), !tv2.equals(zero)) // 7.  tv4 = CMOV(Z, -tv2, tv2 != 0)
+        tv4 = A.mul(tv4) //             8.  tv4 = A * tv4
+        tv2 = tv3.square() //           9.  tv2 = tv3^2
+        let tv6 = tv4.square() //       10. tv6 = tv4^2
+        let tv5 = A.mul(tv6) //         11. tv5 = A * tv6
+        tv2 = tv2.add(tv5) //           12. tv2 = tv2 + tv5
+        tv2 = tv2.mul(tv3) //           13. tv2 = tv2 * tv3
+        tv6 = tv6.mul(tv4) //           14. tv6 = tv6 * tv4
+        tv5 = B.mul(tv6) //             15. tv5 = B * tv6
+        tv2 = tv2.add(tv5) //           16. tv2 = tv2 + tv5
+        let x = tv1.mul(tv3) //         17.   x = tv1 * tv3
+        const { isQR, root: y1 } = sqrt_ratio_3mod4(tv2, tv6) // 18. (is_gx1_square, y1) = sqrt_ratio(tv2, tv6)
+        let y = tv1.mul(u) //           19.   y = tv1 * u
+        y = y.mul(y1) //                20.   y = y * y1
+        x = cmov(x, tv3, isQR) //       21.   x = CMOV(x, tv3, is_gx1_square)
+        y = cmov(y, y1, isQR) //        22.   y = CMOV(y, y1, is_gx1_square)
+        const e1 = sgn(u) === sgn(y) // 23.  e1 = sgn0(u) == sgn0(y)
+        y = cmov(zero.sub(y), y, e1) // 24.   y = CMOV(-y, y, e1)
+        const z = tv4 //                25.   x = x / tv4
+        x = x.mul(z) //                 26. return (x, y, z)
```

Comment:
Just noting that step 26 of https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-14#appendix-F.2 says `return (x, y)` instead of `return (x, y, z)`, but I suppose this is just do to a difference in coordinate systems.

Reply:
Correct!